cmd/k8s-operator/tsrecorder.go

```diff
@@ -0,0 +1,380 @@
+		return nil
+	} else if !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	var sts appsv1.StatefulSet
+	stsKey := types.NamespacedName{
+		Namespace: r.tsNamespace,
+		Name:      tsr.Name,
+	}
+	if err := r.Get(ctx, stsKey, &sts); err != nil && !apierrors.IsNotFound(err) {
+		return err
+	} else if err == nil {
+		// StatefulSet exists, so we have already created the secret.
+		// If the secret is missing, they should delete the TSRecorder CR.
+		logger.Errorf("tsrecorder secret doesn't exist, but the corresponding StatefulSet %s/%s already does. Something is wrong, please delete the TSRecorder CR", sts.GetNamespace(), sts.GetName())
+		return nil
+	}
+	// Create API Key secret which is going to be used by the statefulset
+	// to authenticate with Tailscale.
+	logger.Debugf("creating authkey for new tsrecorder")
+	authKey, err := newAuthKey(ctx, r.tsClient, tsr.Spec.Tags.Stringify())
+	if err != nil {
+		return err
+	}
+
+	mak.Set(&secret.StringData, "authkey", string(authKey))
+
+	logger.Debug("creating a new Secret for the TSRecorder")
+	if err := r.Create(ctx, secret); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *TSRecorderReconciler) validate(tsr *tsapi.TSRecorder) error {
+	// TODO(tomhjp): Error if multiple storage destinations specified.
+	if tsr.Spec.Storage.Disk.Path == "" {
+		return fmt.Errorf("TSRecorder CR %s must specify a storage destination for recordings", tsr.Name)
+	}
+
+	return nil
+}
+
+func (r *TSRecorderReconciler) getNodeID(ctx context.Context, tsrName string) (id tailcfg.StableNodeID, ok bool, err error) {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: r.tsNamespace,
+			Name:      fmt.Sprintf("%s-0", tsrName),
+		},
+	}
+	if err := r.Get(ctx, client.ObjectKeyFromObject(secret), secret); err != nil {
+		if apierrors.IsNotFound(err) {
+			return "", false, nil
+		}
+
+		return "", false, err
+	}
+
+	// TODO(tomhjp): Should maybe use ipn to parse the following info instead.
+	const currentProfileKey = "_current-profile"
+
+	currentProfile, ok := secret.Data[currentProfileKey]
+	if !ok {
+		return "", false, nil
+	}
+	profileBytes, ok := secret.Data[string(currentProfile)]
+	if !ok {
+		return "", false, nil
+	}
+	var profile profile
+	if err := json.Unmarshal(profileBytes, &profile); err != nil {
+		return "", false, fmt.Errorf("failed to extract node profile info from secret %s: %w", secret.Name, err)
+	}
+
+	ok = profile.Config.NodeID != ""
+	return tailcfg.StableNodeID(profile.Config.NodeID), ok, nil
+}
+
+func (r *TSRecorderReconciler) getDeviceInfo(ctx context.Context, tsrName string) (hostname string, ips []string, ok bool, err error) {
+	nodeID, ok, err := r.getNodeID(ctx, tsrName)
+	if !ok || err != nil {
+		return "", nil, false, err
+	}
+
+	// TODO(tomhjp): The profile info doesn't include addresses, which is why we
+	// need the API. Should we instead update the profile to include addresses?
```

Comment:
I guess they could go under `Config` which contains 'generated' data. I would ask around if there is any reason why they weren't put in the profile in the first place

Reply:
I went down the rabbit hole a bit on this one, I think I'll come back to it in a follow-up PR